### PR TITLE
Change permission before removing temp dir

### LINKED ISSUE #64: Files not cleaned correctly
The created folders are not cleaned completely because of some files which rights don't allow opening them.

### PR DESCRIPTION
**Pull request by musikid**
Close #64

### CODE REVIEW

File: rust/src/runner/context.rs [hunk 1]
```diff
@@ -19,7 +19,7 @@ use strum_macros::EnumIter;
 use tempfile::{tempdir_in, TempDir};
 use thiserror::Error;
 
-use crate::{config::SettingsConfig, test::TestError};
+use crate::{config::SettingsConfig, test::TestError, tests::lchmod};
```
Comment (asomers) [inline rust/src/runner/context.rs:22]: This creates a circular modules use graph.  That's confusing.  If lchmod needs to be used both here and in the tests, I suggest you put it in a "utils" module or something.

File: rust/src/runner/context.rs [hunk 2]
```diff
@@ -228,3 +228,20 @@ impl TestContext {
         thread::sleep(self.naptime)
     }
 }
+
+// We implement Drop to circumvent the errors from unlinking a directory for which
+// search or write permission is denied.
+impl Drop for TestContext {
+    fn drop(&mut self) {
+        let iter = walkdir::WalkDir::new(self.base_path()).into_iter();
```
Comment (asomers) [inline rust/src/runner/context.rs:236]: Is this more or less what `/bin/rm -r` does?

File: rust/src/runner/context.rs [hunk 3]
```diff
@@ -228,3 +228,20 @@ impl TestContext {
...
+        for entry in iter {
+            let entry = match entry {
+                Ok(e) => e,
+                Err(_) => continue,
+            };
+
+            let write_perm = Mode::from_bits_truncate(0o700);
+            let _ = lchmod(entry.path(), write_perm);
```
Comment (asomers) [inline rust/src/runner/context.rs:244]: This is probably slower than first `stat`ing the file to check it for writability.  But probably fast enough.  A bigger concern is for the ENOSPC tests.  With ZFS, it's possible that the chmod operation could fail.

File: rust/src/runner/context.rs (hunk 2 above)
Comment (musikid) [reply rust/src/runner/context.rs:236]: It doesn't remove, but it changes the permissions to allow the destructor of `TempDir` to succeed.

Comment (musikid) [reply rust/src/runner/context.rs:236]: We could maybe write our own `TempDir`?

File: rust/src/runner/context.rs (hunk 1 above)
Comment (musikid) [reply rust/src/runner/context.rs:22]: Oh yes, sorry!

File: rust/src/runner/context.rs (hunk 2 above)
Comment (asomers) [reply rust/src/runner/context.rs:236]: This is equivalent to `chmod -R 700`, which is much more work than necessary.  You should actually only need to change the permissions of directories.  Files can be unlinked even without write permission.  The returned `entry` caches the file type, so you should check that, and skip the chmod if it isn't a directory.
You may also need to `chflags 0` some files before deleting them, if their SF_IMMUTABLE or UF_IMMUTABLE bit is set.

File: rust/src/runner/context.rs (hunk 3 above)
Comment (musikid) [reply rust/src/runner/context.rs]: It seems that these tests don't change the permissions, however. I'm unsure of how we should handle errors in the destructor. Maybe print a warning for the user when it fails?

File: rust/src/runner/context.rs (hunk 2 above)
Comment (musikid) [reply rust/src/runner/context.rs:247]: Thanks, I added the checks (for stat and file type), and `chflags` to reset the flags.

File: rust/src/runner/context.rs [hunk 4]
```diff
@@ -228,3 +239,38 @@ impl TestContext {
...
+                Ok(e) => e,
+                _ => continue,
+            };
+
+            let file_stat = match stat(entry.path()) {
+                Ok(s) => s,
+                _ => continue,
+            };
+
+            if entry.file_type().is_dir() && (file_stat.st_mode & Mode::S_IRWXU.bits()) == 0 {
```
Comment (asomers) [inline rust/src/runner/context.rs:259]: You could save a syscall by doing the file_type() check before calling stat().

File: rust/src/runner/context.rs (hunk 3 above)
Comment (asomers) [reply rust/src/runner/context.rs]: There isn't a whole lot that you can do there.  I think it's ok as is, but you could also print a warning.  The important thing is that you don't panic if the thread is already handling a panic.

File: rust/src/runner/context.rs (hunk 4 above)
Comment (musikid) [reply rust/src/runner/context.rs:259]: It also uses the `stat()` result to check if a flag is set to avoid calling `chflags` for nothing.

Comment (asomers) [reply rust/src/runner/context.rs:259]: Well yeah.  You would still need to call stat if the direntry isn't a directory.   I'm just suggesting you check that it's a directory first.  In that case you wouldn't need to stat it.

File: rust/src/runner/context.rs (hunk 3 above)
Comment (musikid) [reply rust/src/runner/context.rs]: Alright, I will keep it as it then.

**Review by asomers — APPROVED** [review]

**Pull request merged**